main/src/library/flix/core/RedBlackTree.flix

```diff
@@ -0,0 +1,227 @@
+        case Leaf => Leaf
+        case Node(Red, Leaf, y, Leaf) if x == y => Leaf
+        case Node(Red, Leaf, y, Leaf) if x != y => tree
+        case Node(Black, Leaf, y, Leaf) if x == y => DoubleLeaf
+        case Node(Black, Leaf, y, Leaf) if x != y => tree
+        case Node(Black, Node(Red, Leaf, y, Leaf), z, Leaf) if x < z =>
+            Node(Black, deleteHelper(x, Node(Red, Leaf, y, Leaf)), z, Leaf)
+        case Node(Black, Node(Red, Leaf, y, Leaf), z, Leaf) if x == z =>
+            Node(Black, Leaf, y, Leaf)
+        case Node(Black, Node(Red, Leaf, y, Leaf), z, Leaf) if x > z =>
+            Node(Black, Node(Red, Leaf, y, Leaf), z, Leaf)
+        case Node(c, a, y, b) if x < y => rotate(Node(c, deleteHelper(x, a), y, b))
+        case Node(c, a, y, b) if x == y =>
+            let (z, e) = minDelete(b);
+            rotate(Node(c, a, z, e))
+        case Node(c, a, y, b) if x > y => rotate(Node(c, a, y, deleteHelper(x, b)))
+        case _ => tree
+    }
+
+    ///
+    /// Returns a new, balanced tree containg all elements of `tree` execept `x`.
+    ///
+    pub def delete(x: a, tree: RedBlackTree[a]): RedBlackTree[a] = redden(deleteHelper(x, tree))
+
+    ///
+    /// Returns true if and only if `tree` contains `x`.
+    ///
+    pub def contains(x: a, tree: RedBlackTree[a]): Bool = match tree {
+        case Leaf => false
+        case Node(_, a, y, _) if x < y => contains(x, a)
+        case Node(_, _, y, b) if x > y => contains(x, b)
+        case _ => true
+    }
+
+    ///
+    /// Returns the height of `tree`.
+    ///
+    pub def height(tree: RedBlackTree[a]): Int32 = match tree {
+        case Node(_, a, _, b) => if (height(a) > height(b)) 1 + height(a) else 1 + height(b)
+        case _ => 0
+    }
+
+    ///
+    /// Returns the number of nodes in `tree`.
+    ///
+    pub def size(tree: RedBlackTree[a]): Int32 = match tree {
+        case Node(_, a, _, b) => 1 + size(a) + size(b)
+        case _ => 0
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `tree` going from left to right, i.e. inorder.
+    ///
+    ///
+    pub def foldLeft(f: (b, a) -> b, s: b, tree: RedBlackTree[a]): b = match tree {
+        case Node(_, a, x, b) => foldLeft(f, f(foldLeft(f, s, a), x), b)
+        case _ => s
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `tree` going from right to left, i.e. reverse inorder
+    ///
+    pub def foldRight(f: (b, a) -> b, s: b, tree: RedBlackTree[a]): b = match tree {
+        case Node(_, a, x, b) => foldRight(f, f(foldRight(f, s, b), x), a)
+        case _ => s
+    }
+
+    ///
+    /// Optionally returns the left most element of `tree`.
+    ///
+    pub def leftMost(tree: RedBlackTree[a]): Option[a] = match tree {
+        case Node(_, a, x, _) => match leftMost(a) {
+            case None => Some(x)
+            case _ => leftMost(a)
+        }
+        case _ => None
```

Comment:
I would think that the above pattern match can be significantly simplified. At least such that it is proper tail recursive.